WalletWasabi.Packager/MacSignTools.cs

```diff
@@ -0,0 +1,394 @@
+
+			UnlockKeychain();
+
+			Sign($"{signArguments} --entitlements \"{entitlementsPath}\" \"{dmgFilePath}\"", dmgPath);
+
+			Console.WriteLine("Phase: verifying the signature.");
+
+			Verify(dmgFilePath);
+
+			Console.WriteLine("Phase: notarize dmg");
+			Notarize(appleId, password, dmgFilePath, bundleIdentifier);
+
+			Console.WriteLine("Phase: staple dmp");
+			Staple(dmgFilePath);
+
+			File.Move(dmgFilePath, desktopDmgFilePath);
+			DeleteWithChmod(workingDir);
+			File.Delete(zipPath);
+
+			Console.WriteLine("Phase: finish.");
+		}
+
+		public static bool IsMacSignMode(string[] args)
+		{
+			return RuntimeInformation.IsOSPlatform(OSPlatform.OSX); // For now this is enough. If you run it on macOS you want to sign.
+		}
+
+		private static void Notarize(string appleId, string password, string filePath, string bundleIdentifier)
+		{
+			string uploadId = null;
+
+			Console.WriteLine("Start notarizing, uploading file.");
+
+			using (var process = Process.Start(new ProcessStartInfo
+			{
+				FileName = "xcrun",
+				Arguments = $"altool --notarize-app -t osx -f \"{filePath}\" --primary-bundle-id \"{bundleIdentifier}\" -u \"{appleId}\" -p \"{password}\" --output-format xml",
+				RedirectStandardOutput = true,
+			}))
+			{
+				process.WaitForExit();
+				string result = process.StandardOutput.ReadToEnd();
+
+				if (result.Contains("The software asset has already been uploaded. The upload ID is"))
+				{
+					// Example: The software asset has already been uploaded. The upload ID is 7689dc08-d6c8-4783-8d28-33e575f5c967
+					uploadId = result.Split('"').First(line => line.Contains("The software asset has already been uploaded.")).Split("The upload ID is").Last().Trim();
+				}
+				else if (result.Contains("No errors uploading"))
+				{
+					// Example: <key>RequestUUID</key>\n\t\t<string>2a2a164f-2ae7-4293-8357-5d5a5cdd580a</string>
+
+					var lines = result.Split('\n');
+
+					for (int i = 0; i < lines.Length; i++)
+					{
+						string line = lines[i].Trim();
+						if (!line.TrimStart().StartsWith("<key>", StringComparison.InvariantCultureIgnoreCase))
+						{
+							continue;
+						}
+
+						if (line.Contains("<key>RequestUUID</key>", StringComparison.InvariantCulture))
+						{
+							uploadId = lines[i + 1].Trim().Replace("<string>", "").Replace("</string>", "");
+						}
+					}
+				}
+			}
+
+			if (uploadId is null)
+			{
+				throw new InvalidOperationException("Cannot get uploadId. Notarization failed.");
+			}
+
+			Stopwatch sw = new Stopwatch();
+			sw.Start();
+			while (true) // Wait for the notarization.
+			{
+				Console.WriteLine($"Checking notarization status. Elapsed time: {sw.Elapsed.ToString()}");
+				using var process = Process.Start(new ProcessStartInfo
+				{
+					FileName = "xcrun",
+					Arguments = $"altool --notarization-info \"{uploadId}\" -u \"{appleId}\" -p \"{password}\"",
+					RedirectStandardError = true,
+				});
+				process.WaitForExit();
```

Comment:
Here again I would check the `ErrorCode`. I know we are checking the standard output but if there is a problem it doesn't make sense to wait those 4 seconds.